### PR TITLE
supervisor-upgrades: limit to devices that support multicontainer

### PR DESCRIPTION
**Pull request by xginn8**
Connects-to: https://github.com/balena-io/balena-sdk/issues/1033
Change-type: patch
Signed-off-by: Matthew McGinn <matthew@balena.io>

Resolves: # 
HQ: <url> 
See: <url> 
Depends-on: <url> 
Change-type: major|minor|patch 

---
##### Contributor checklist

- [ ] Includes tests
- [ ] Includes typings
- [ ] Includes updated documentation
- [ ] Includes updated build output

### CODE REVIEW

**Review by sradevski — APPROVED** [review]

File: tests/integration/models/device.spec.js [hunk 1]
```diff
@@ -2725,6 +2728,25 @@ describe('Device Model', function () {
 						.expect(promise)
 						.to.be.rejectedWith(`Release not found: ${badRelease}`);
 				});
+
+				describe('given a device that does not support multicontainer', function () {
```
Comment (thgreasi) [inline tests/integration/models/device.spec.js:2732]: Maybe we should wrap the old tests in a `given a device that does support multicontainer`, since both the `givenADevice(beforeEach` of that `describe()` & the `givenADevice(before` in here run for this test case.
What do you think @xginn8 ?

File: lib/models/device.ts [hunk 2]
```diff
@@ -2998,6 +3005,25 @@ const getDeviceModel = function (
...
+			const device = (await exports.get(
+				uuidOrId,
+				deviceOpts,
+			)) as PineTypedResult<Device, typeof deviceOpts>;
+			ensureVersionCompatibility(
+				device.supervisor_version,
+				MIN_SUPERVISOR_MC_API,
+				'supervisor',
+			);
+			ensureVersionCompatibility(device.os_version, MIN_OS_MC, 'host OS');
```
Comment (thgreasi) [inline lib/models/device.ts:3026]: Does the API do any such check as well?
I'm not against this, since it gives a better error message, but having this in a hook of the API might be better.

File: tests/integration/models/device.spec.js [hunk 3]
```diff
@@ -2679,6 +2679,9 @@ describe('Device Model', function () {
 			});
 
 			describe('balena.models.device.setSupervisorRelease()', function () {
+				givenADevice(beforeEach, {
+					os_version: 'balenaOS 2.56.1+rev1',
+				});
```
Comment (thgreasi) [inline tests/integration/models/device.spec.js:2684]: NIT: How about just patching the device's os_version instead of creating a new one (on top of the `given a newly registered offline device` adds in its own `beforeEach`?
(Same below)

Comment (xginn8) [reply tests/integration/models/device.spec.js:2684]: sure, i avoided that as there may be side effects we are bypassing by faking a HUP.

File: lib/models/device.ts (hunk 2 above)
Comment (xginn8) [reply lib/models/device.ts:3026]: the API does not enforce this at the moment, as it felt a little overkill for a small edge case. i can enforce it as part of a hook if you'd be more comfortable enforcing it in both places.

File: tests/integration/models/device.spec.js (hunk 3 above)
Comment (thgreasi) [reply tests/integration/models/device.spec.js:2684]: I'm fine if you prefer to do it like this (a POST and not a PATCH), I'm only trying to save the duplicate device creation in before each test case.
Feel free to move the whole `describe('balena.models.device.setSupervisorRelease()', function () {` outside the `'given a newly registered offline device'` and just have the `beforeEach` exactly as you prefer it :)

File: tests/integration/models/device.spec.js (hunk 1 above)
Comment (xginn8) [reply tests/integration/models/device.spec.js:2735]: done, good idea.

File: tests/integration/models/device.spec.js (hunk 3 above)
Comment (xginn8) [reply tests/integration/models/device.spec.js]: done, thanks for the rec!

File: lib/models/device.ts (hunk 2 above)
Comment (xginn8) [reply lib/models/device.ts:3026]: thinking through this, we can't enforce this at the API level since devices that are being HUPed still require supervisor updates. we could potentially limit it to only device API keys, but i'm sure if that complexity is worth it.

**Review by thgreasi — COMMENTED** [review]
I probably failed to describe this properly.
But I pushed two commits on this branch.
Let me know what you think, squash the typo commit and self-certify.

**Review by thgreasi — APPROVED** [review]

**Comment by xginn8** [comment]
@balena-ci retest

**Comment by thgreasi** [comment]
@balena-ci retest
